### PR TITLE
udp: set Don't Fragment(DF) bit in IP packet header

### PR DESCRIPTION
**Pull request by danzh2010**
Commit Message: set corresponding socket options to set do not fragment bit in IP packet header for UDP listener sockets and QUIC upstream connection sockets.

Additional Description: also fix getLoopbackAddress() in envoy_quic_utils.cc to use the same socket interface and v6_only bit as peer_address.

Risk Level: medium, change the DF bit on the wire
Testing: new unit tests
Docs Changes: N/A
Release Notes: Yes
Platform Specific Features: N/A
Runtime guard: envoy.reloadable_features.udp_set_do_not_fragment

### CODE REVIEW

File: source/common/quic/envoy_quic_utils.cc [hunk 1]
```diff
@@ -215,6 +217,25 @@ createConnectionSocket(const Network::Address::InstanceConstSharedPtr& peer_addr
...
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.udp_set_do_not_fragment")) {
+    int v6_only = 0;
+    if (connection_socket->ipVersion().has_value() &&
+        connection_socket->ipVersion().value() == Network::Address::IpVersion::v6) {
+      socklen_t v6_only_len = sizeof(v6_only);
+      Api::SysCallIntResult result =
+          connection_socket->getSocketOption(IPPROTO_IPV6, IPV6_V6ONLY, &v6_only, &v6_only_len);
+      if (result.return_value_ != 0) {
+        ENVOY_LOG_MISC(error,
+                       "Fail to get IPV6_V6ONLY socket option, getsockopt() return {}, errno {}",
```
Comment (RyanTheOptimist) [inline source/common/quic/envoy_quic_utils.cc:229]: nit: "Fail" => "Failed" and "return" => "returned", I think

File: source/common/quic/envoy_quic_utils.cc [hunk 2]
```diff
@@ -215,6 +217,25 @@ createConnectionSocket(const Network::Address::InstanceConstSharedPtr& peer_addr
   if (prefer_gro && Api::OsSysCallsSingleton::get().supportsUdpGro()) {
     connection_socket->addOptions(Network::SocketOptionFactory::buildUdpGroOptions());
   }
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.udp_set_do_not_fragment")) {
+    int v6_only = 0;
+    if (connection_socket->ipVersion().has_value() &&
+        connection_socket->ipVersion().value() == Network::Address::IpVersion::v6) {
```
Comment (RyanTheOptimist) [inline source/common/quic/envoy_quic_utils.cc:223]: In the case of the listener socket up above, it looks like we don't call getsocketopt(). What's the difference which requires us to do so here?

**Review by RyanTheOptimist — COMMENTED** [review]
Exciting!

File: source/common/quic/envoy_quic_utils.cc (hunk 2 above)
Comment (danzh2010) [reply source/common/quic/envoy_quic_utils.cc:223]: Whether a listen socket is v6 only or not purely depends on the configured listening address. If the address is v6 only, Envoy will call setsockopt() with V6ONLY option. But client connection sockets in addition to the configured address, also depends on platform (Android and iOS don't set the socket option V6ONLY even if the address is configured v6 only), so checking the option value here is more straight forward for client sockets.

File: source/common/network/socket_option_factory.h [hunk 3]
```diff
@@ -38,6 +38,10 @@ class SocketOptionFactory : Logger::Loggable<Logger::Id::connection> {
   static std::unique_ptr<Socket::Options> buildUdpGroOptions();
   static std::unique_ptr<Socket::Options> buildZeroSoLingerOptions();
   static std::unique_ptr<Socket::Options> buildIpRecvTosOptions();
+  /**
+   * @param mapped_v6 true if this option is to be applied to a dual socket.
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/common/network/socket_option_factory.h:42]: We only want to apply the additional ENVOY_IP_MTU_DISCOVER option to v6 socket that support v4 mapped v6 address. So this param should indicate v6 socket AND not v6_only. I find the name v6_only isn't sufficient to convey this information.

**Review by RyanTheOptimist — COMMENTED** [review]
Looking good!

File: source/common/network/socket_option_factory.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply source/common/network/socket_option_factory.h:42]: ARgh! Github lots my comment somehow. I'm fine with either name, but lets' be consistent. Right now the comment says "dual stack" but the parameter says "mapped v6". Let's use the same terminology to be consistent and make sure the comment has enough detail. I might suggests "supports_v4_mapped_v6_addresses" as that seems to be the semantics we really want here?

**Review by RyanTheOptimist — DISMISSED** [review]

**Comment by danzh2010** [comment]
Fixed test coverage. PTAL!

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**